### PR TITLE
feature:State machine common package + Node state service

### PR DESCRIPTION
**Pull request by Brackleycassinga**
This pull request introduces a state management system for nodes, enhancing the handling of node state transitions and event processing.

**Changes Made:**

- State Common Package: Introduced a common package for managing state-related operations.
- Node State Service: Implemented a dedicated state service under the node management module.
- Event Handling: Added functionality to effectively manage and respond to events for nodes.
- StateEventServer Implementation: Developed the StateEventServer to facilitate node state transitions.
- Event Processing and Buffering: Implemented methods to process incoming events and maintain an event buffer.
- StateMachineInstance Integration: Integrated StateMachineInstance to oversee the node lifecycle and transitions.
- Message Bus Publishing: Added the capability to publish state changes via the message bus for improved communication.
- Thread-Safe Operations: Ensured thread-safe handling of state instances and event buffers to maintain data integrity in concurrent environments.

### CODE REVIEW

File: systems/common/events/node_state_event.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		Key:        NodeStateEventConfig,
+		Name:       "config",
+		RoutingKey: NodeStateEventRoutingKey[NodeStateEventConfig],
+		body:       pb.NodeFeederMessage{},
+	},
+	NodeStateEventReady: {
+		Key:        NodeStateEventReady,
+		Name:       "ready",
+		RoutingKey: NodeStateEventRoutingKey[NodeStateEventReady],
+		// body:epb.NodeReadyEvent{},
```
Comment (delvatt) [inline systems/common/events/node_state_event.go:86]: Do we need body? If not, shall we remove?

**Review by delvatt — APPROVED** [review]
LGTM! although, I'm not very acquainted with the full design therefore I can only provide a high level review, so: 

- better have another reviewer who is more familiar with the whole design. 
-  make sure you remove commented code if no more useful

File: systems/common/stateMachine/state_machine.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+	if currentState.OnExit != nil {
+		if err := currentState.OnExit(); err != nil {
+			return fmt.Errorf("error in OnExit for state %s: %w", instance.CurrentState, err)
+		}
+	}
+
+	log.Infof("Current state: %s", currentState.Name)
+
+	if transition, exists := currentState.Transitions[eventName]; exists {
```
Comment (vthakur7f) [inline systems/common/stateMachine/state_machine.go:182]: As discussed we need to make sure transitions from the sub-state to next state are happening.

File: systems/node/state/pkg/config.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+func NewConfig(name string) *Config {
+	return &Config{
+		DB: &uconf.Database{
+			DbName: name,
+		},
+		Service: uconf.LoadServiceHostConfig(name),
+		MsgClient: &uconf.MsgClient{
+			Timeout: 7 * time.Second,
+			ListenerRoutes: []string{
+				evt.NodeStateEventRoutingKey[evt.NodeStateEventAssign],
```
Comment (vthakur7f) [inline systems/node/state/pkg/config.go:43]: Keys need to be updated based on the notify service

**Review by vthakur7f — COMMENTED** [review]
Good work. Lets' compelete it :-)

**Review by vthakur7f — COMMENTED** [review]
Good work. Lets' compelete it :-)

File: systems/node/state/pb/state.proto [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+import "ukama/node.proto";
+
+service StateService {
+    rpc AddState (AddStateRequest) returns (AddStateResponse);
+    rpc GetStateById (GetStateByIdRequest) returns (GetStateByIdResponse);
+    rpc GetStates(GetStatesRequest) returns(GetStatesResponse);
+    rpc GetLatestState(GetLatestStateRequest) returns (GetLatestStateResponse);  
+    rpc UpdateState(UpdateStateRequest) returns (UpdateStateResponse);
```
Comment (vthakur7f) [inline systems/node/state/pb/state.proto:21]: We need to get Node State History (could be based on a time filter or you could be just last N number of states)

File: systems/common/stateMachine/state_machine.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+	cache.mu.RUnlock()
+
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if config, exists := cache.configs[configFile]; exists {
+		return config, nil
+	}
+
+	data, err := ioutil.ReadFile(configFile)
```
Comment (vthakur7f) [inline systems/common/stateMachine/state_machine.go:375]: use os.ReadFile instead of ioutil

**Review by vthakur7f — APPROVED** [review]
Some work is still left in the state machine. May be it's a good idea to document them as new issues.

**Review by vthakur7f — APPROVED** [review]
Can you resolve all the old conversatons

**Pull request merged**